compliance/collection/main.go

```diff
@@ -182,17 +217,106 @@ func manageNodeScanLoop(ctx context.Context, rescanInterval time.Duration, scann
 	return sensorC
 }
 
-func scanNode(nodeName string, scanner nodeinventorizer.NodeInventorizer) (*sensor.MsgFromCompliance, error) {
-	result, err := scanner.Scan(nodeName)
+func runCachedScan(opts *cachedScanOpts) (*storage.NodeInventory, error) {
+	inventory, err := opts.scanner.Scan(opts.nodeName)
 	if err != nil {
 		return nil, err
 	}
+	inv, err := proto.Marshal(inventory)
+	if err != nil {
+		return nil, err
+	}
+	if err := os.WriteFile(fmt.Sprintf("%s/last_scan", opts.inventoryCachePath), inv, 0600); err != nil {
+		return nil, err
+	}
+
+	return inventory, nil
+}
+
+func createAndObserveMessage(nodeName string, inventory *storage.NodeInventory) *sensor.MsgFromCompliance {
 	msg := &sensor.MsgFromCompliance{
 		Node: nodeName,
-		Msg:  &sensor.MsgFromCompliance_NodeInventory{NodeInventory: result},
+		Msg:  &sensor.MsgFromCompliance_NodeInventory{NodeInventory: inventory},
 	}
 	cmetrics.ObserveInventoryProtobufMessage(msg)
-	return msg, nil
+	return msg
+}
+
+// scanNodeWithBackoff runs scans with a linear backoff based on a file to not overstrain a Node if the container keeps restarting.
+// The backoff file will only be encountered if the previous container is killed during a call to scanNodeWithBackoff.
+// Note: This does not prevent strain in case of repeated pod recreation, as it is based on an EmptyDir.
+func scanNodeWithBackoff(opts *cachedScanOpts) (*sensor.MsgFromCompliance, error) {
+	backoffIntervalSeconds := int64(env.NodeInventoryInitialBackoff.DurationSetting() / time.Second)
+
+	backoffFile, err := os.ReadFile(fmt.Sprintf("%s/backoff", opts.inventoryCachePath))
+	defer func() {
+		if err := os.Remove(fmt.Sprintf("%s/backoff", opts.inventoryCachePath)); err != nil {
+			log.Warnf("Could not remove backoff state file: %v", err)
```

Comment:
1. Please don't Warn if the file does not exist.
2. Suggestion:
```suggestion
			log.Warnf("Could not remove scan backoff state file: %v", err)
```

Reply:
I've updated the log message, thanks for that, but I don't understand your first point. 
There is no warning when the file doesn't exist. We have a message for that case, but it is on Debug level (line 2. Is this what you meant?
https://github.com/stackrox/stackrox/blob/53f171bbbe7430bd09bccd8fc6c5178d2f04e280/compliance/collection/main.go#L259

Reply:
At that state of the code, the function that wrote backoff file may have failed to create it, but we still try to delete the file and log error because it should also fail due to a file not being there.
This isn't any more a concern in the current state of the code.